### PR TITLE
new: metadata kind email and new control for Audit

### LINKED ISSUE #164: Support for timing audit
opendatakit/xforms-spec#94

The agreed-upon XForms way of enabling a timing audit:
```xml
<orx:meta>
    <orx:audit/>
</orx:meta>

<bind nodeset="/orx:meta/orx:audit" type="binary"/>
```

For now, including this would mean a csv file with detailed timing information is included with each form submission. In the future, there will also be the option to customize the contents of the csv (e.g. add location information) through attributes.

More about the feature and how it works is at opendatakit/collect#257

There isn't yet an agreed-upon XLSForm syntax - XLSForm/pyxform#128

### LINKED ISSUE #209: Support all metadata kinds: email, phoneNumber, audit, etc
In addition to DeviceID, add Userid to metadata. Some discussion here: https://forum.opendatakit.org/t/collecting-user-email-address-in-form-metadata/15686

### PR DESCRIPTION
**Pull request by florianm**
This PR brings the two final new features of frankenPR #266 (fixes #266) and fixes #164 and fixes #209.

In a small addition, the last missing metadata kind "email" is added in two places:
* 1/2: email is added to the list of available metadata kinds,
* 2/2: email is added to the XML in the same way as e.g. username.
I've included this here as the code changes are closely related for review to require the same brainspace.

The main part of the PR is to support audit, which is different to the other metadata fields as it has its own [four attributes](https://getodk.github.io/xforms-spec/#audit-attributes), but always has the label "Audit" (so that can be made implicit).
Audit is added as a top level control element. Much of the handling is similar to metadata (there, some `if` clauses get extended to include the control type `audit`), some is different enough to warrant its own code block. I'll annotate the code changes with what the sections do.

## Walkthrough
Let's build a simple form to show the new behaviour of email and audit.
One text field:
image

The required namespace `orx` was already added in #281.
We now see the meta node being prefixed as `orx:meta` as per the [XForms metadata spec](https://getodk.github.io/xforms-spec/#metadata).

```
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:jr="http://openrosa.org/javarosa" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:orx="http://openrosa.org/xforms" xmlns:odk="http://www.opendatakit.org/xforms">
  <h:head>
    <h:title>164 Audit Email</h:title>
    <model>
      <instance>
        <data id="164-Audit-Email" orx:version="1643605416">
          <orx:meta>
            <orx:instanceID/>
          </orx:meta>
          <text1/>
        </data>
      </instance>
      <itext>
        <translation lang="English">
          <text id="/data/text1:label">
            <value>Text 1 label</value>
          </text>
          <text id="/data/text1:hint">
            <value>Text 1 hint</value>
          </text>
        </translation>
      </itext>
      <bind nodeset="/data/meta/instanceID" type="string" readonly="true()" jr:preload="uid"/>
      <bind nodeset="/data/text1" type="string"/>
    </model>
  </h:head>
  <h:body>
    <input ref="/data/text1">
      <label ref="jr:itext('/data/text1:label')"/>
      <hint ref="jr:itext('/data/text1:hint')"/>
    </input>
  </h:body>
</h:html>
```

Adding existing metadata `username` and new metadata `email`, we see that email behaves exactly like username:
image

```
      <bind nodeset="/data/meta/instanceID" type="string" readonly="true()" jr:preload="uid"/>
      <bind nodeset="/data/text1" type="string"/>
      <bind nodeset="/data/username" type="string" jr:preload="property" jr:preloadParams="username"/>
      <bind nodeset="/data/email" type="string" jr:preload="property" jr:preloadParams="email"/>
```

Adding a new control "Audit" shows the audit attributes as first level properties, populated with defaults, and the control being labelled as "Audit". The data name also defaults to "audit" but can be adjusted.
image
The XML output now reads:
```
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:jr="http://openrosa.org/javarosa" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:orx="http://openrosa.org/xforms" xmlns:odk="http://www.opendatakit.org/xforms">
  <h:head>
    <h:title>164 Audit Email</h:title>
    <model>
      <instance>
        <data id="164-Audit-Email" orx:version="1643605882">
          <orx:meta>
            <orx:instanceID/>
            <orx:audit/>
          </orx:meta>
          <text1/>
          <username/>
          <email/>
        </data>
      </instance>
      <itext>
        <translation lang="English">
          <text id="/data/text1:label">
            <value>Text 1 label</value>
          </text>
          <text id="/data/text1:hint">
            <value>Text 1 hint</value>
          </text>
        </translation>
      </itext>
      <bind nodeset="/data/meta/instanceID" type="string" readonly="true()" jr:preload="uid"/>
      <bind nodeset="/data/text1" type="string"/>
      <bind nodeset="/data/username" type="string" jr:preload="property" jr:preloadParams="username"/>
      <bind nodeset="/data/email" type="string" jr:preload="property" jr:preloadParams="email"/>
      <bind nodeset="/data/meta/audit" type="binary" odk:location-priority="balanced" odk:location-min-interval="20" odk:location-max-age="60" odk:track-changes="true"/>
    </model>
  </h:head>
  <h:body>
    <input ref="/data/text1">
      <label ref="jr:itext('/data/text1:label')"/>
      <hint ref="jr:itext('/data/text1:hint')"/>
    </input>
  </h:body>
</h:html>
```

Changing the audit parameter `location priority` to "off" creates the bind node 
```
<bind nodeset="/data/meta/audit" type="binary" odk:track-changes="true"/>
```

Changing the audit parameter `track changes` to "false" creates the bind node
```
      <bind nodeset="/data/meta/audit" type="binary" odk:track-changes="false"/>
```
which obviously makes sense only if combined with location tracking enabled (prio other than "off"), e.g.
```
      <bind nodeset="/data/meta/audit" type="binary" odk:location-priority="high-accuracy" odk:location-min-interval="20" odk:location-max-age="60" odk:track-changes="false"/>
```

## Form and submissions
[164-Audit-Email.zip](https://github.com/getodk/build/files/7968320/164-Audit-Email.zip)
The attached ZIP is the export of two test submissions to the (also included) XForm. I've edited out my (verified correct) coordinates Ith `xxx` where recorded by the audit log. I can also see my email which I've set in the Collect metadata as "noreply@email.org" - looks like a default, but that was actually me. The top level audit log contains the audit for both submissions, the media/audit.csv contains only the last form's audit log.

From these test submissions I conclude that the email and audit controls work and generate a valid XForm. I didn't test any edge cases or other forms yet.

### CODE REVIEW

File: public/javascripts/control.js [hunk 1]
```diff
@@ -92,6 +92,10 @@
         {
             $headline.children('.controlLabel').text(properties.kind.value);
         }
+        else if (type == 'audit')
+        {
+            $headline.children('.controlLabel').text("Audit");
+        }
```
Comment (florianm) [inline public/javascripts/control.js:98]: Audit controls get a fixed label "Audit". Users can choose a separate data name (default: "audit" instead of "untitled").

File: public/javascripts/control.js [hunk 2]
```diff
@@ -333,7 +337,7 @@
 
             if (defaultProperties != null)
                 properties = defaultProperties;
-            else if ((type == 'group') || (type == 'branch') || (type == 'metadata'))
+            else if ((type == 'group') || (type == 'branch') || (type == 'metadata') || (type == 'audit'))
```
Comment (florianm) [inline public/javascripts/control.js:340]: Like metadata, audit does not get default control properties (label, help etc).

File: public/javascripts/control.js [hunk 3]
```diff
@@ -836,9 +840,62 @@
           kind:       { name: 'Kind',
                         type: 'enum',
                         description: 'Type of metadata to add.',
-                        options: [ 'Device ID', 'Start Time', 'End Time', 'Today', 'Username', 'Subscriber ID', 'SIM Serial', 'Phone Number', 'Start Geopoint' ],
+                        options: [ 'Device ID', 'Start Time', 'End Time', 'Today', 'Username', 'Email', 'Subscriber ID', 'SIM Serial', 'Phone Number', 'Start Geopoint' ],
```
Comment (florianm) [inline public/javascripts/control.js:843]: Email 1/2: add email as available metadata kind.

File: public/javascripts/control.js [hunk 4]
```diff
@@ -836,9 +840,62 @@
...
+                          summary: false },
+            track_changes: {
+                          name: 'Track Changes',
+                          type: 'enum',
+                          description: 'Whether to track changes to the form made any time before submission.',
+                          tips: [],
+                          options: ['true', 'false'],
+                          value: 'true',
+                          summary: false },
+          },
```
Comment (florianm) [inline public/javascripts/control.js:898]: Audit gets a custom set of control properties: no `kind` (as there's only one choice, "audit"), plus the four metadata attributes.

File: public/javascripts/control.js [hunk 5]
```diff
@@ -918,7 +975,11 @@
         },
         metadata: {
             name: 'Metadata',
-            description: 'Metadata questions silently and automatically collect information about the session.',
+            description: 'Metadata questions silently and automatically collect information about the session. '
+        },
+        audit: {
+            name: 'Audit',
+            description: 'Audit metadata can track form changes and/or device location.',
```
Comment (florianm) [inline public/javascripts/control.js:982]: Context help for audit control.

File: public/javascripts/data.js [hunk 6]
```diff
@@ -63,17 +63,25 @@ var dataNS = odkmaker.namespace.load('odkmaker.data');
...
+     * Each time a form is loaded, this function populates the 
+     * "properties" sidebar with the control's properties as defined in 
+     * `$.fn.odkControl.controlProperties` (`control.js`).
+     * 
+     */
     var loadOne = odkmaker.data.loadOne = function(control, $parent)
     {
         var properties = null;
-        if ((control.type == 'group') || (control.type == 'branch') || (control.type == 'metadata'))
+        if ((control.type == 'group') || (control.type == 'branch') || (control.type == 'metadata') || (control.type == 'audit'))
```
Comment (florianm) [inline public/javascripts/data.js:77]: No default properties for audit, like metadata.

File: public/javascripts/data.js [hunk 7]
```diff
@@ -63,17 +63,25 @@ var dataNS = odkmaker.namespace.load('odkmaker.data');
...
             properties = $.extend(true, {}, $.fn.odkControl.controlProperties[control.type]);
         else
             properties = $.extend(true, $.extend(true, {}, $.fn.odkControl.defaultProperties),
                                         $.fn.odkControl.controlProperties[control.type]);
         _.each(properties, function(property, key)
         {
-            if (key === 'metadata') return;
+            if (key === 'metadata' || key === 'audit') return;
             property.value = control[key];
```
Comment (florianm) [inline public/javascripts/data.js:85]: Like metadata, audit captures no user-entered value.

File: public/javascripts/data.js [hunk 8]
```diff
@@ -433,7 +441,66 @@ var dataNS = odkmaker.namespace.load('odkmaker.data');
...
+            };
+
+            if (control.track_changes != undefined) {
+                binding.attrs['odk:track-changes'] = control.track_changes;
+            };
+
+            model.children.push(binding);
+
+            return;
+        }
```
Comment (florianm) [inline public/javascripts/data.js:501]: This pushes two elements into the exported XML, meta and bind node. See inline comments.

File: public/javascripts/data.js [hunk 9]
```diff
@@ -491,6 +558,11 @@ var dataNS = odkmaker.namespace.load('odkmaker.data');
                 binding.attrs['jr:preload'] = 'property';
                 binding.attrs['jr:preloadParams'] = 'username';
             }
+            else if (kind == 'email') {
+                binding.attrs.type = 'string';
+                binding.attrs['jr:preload'] = 'property';
+                binding.attrs['jr:preloadParams'] = 'email';
+            }
```
Comment (florianm) [inline public/javascripts/data.js:565]: Email 2/2: push the bind node into the exported XML.

File: server/views/index.erb [hunk 10]
```diff
@@ -123,6 +123,7 @@
           <li><a class="toolButton inputSelectMany" rel="inputSelectMany">Select Multiple</a></li>
           <li class="separator"></li>
           <li><a class="toolButton metadata" rel="metadata">Metadata</a></li>
+          <li><a class="toolButton audit" rel="audit">Audit</a></li>
```
Comment (florianm) [inline server/views/index.erb:126]: Footer menu: add new control type "Audit". No changes to `styles.css` needed, the class `toolButton` applies the correct style.

**Comment by lognaturel** [comment]
> meta node being prefixed as orx:meta

Note that Enketo doesn't currently support editing records that have a namespaced meta block: https://github.com/enketo/enketo-core/issues/835. This update will be available in Central v1.14 (~2 weeks from now). I think this can move forward anyway but wanted to capture this info.

**Comment by florianm** [comment]
> > meta node being prefixed as orx:meta
> 
> Note that Enketo doesn't currently support editing records that have a namespaced meta block: [enketo/enketo-core#835](https://github.com/enketo/enketo-core/issues/835). This update will be available in Central v1.14 (~2 weeks from now). I think this can move forward anyway but wanted to capture this info.

Good to know! I just followed the XForms spec here. Seeing the Enketo fix is rolled out by the time we've got Build and build2xlsform PRs merged and deployed, do I understand you correctly to keep the `orx` prefixes?

**Comment by lognaturel** [comment]
> The data name also defaults to "audit" but can be adjusted.

The audit is only populated if there's a node named `audit` in the model so this can't be configurable.

**Comment by lognaturel** [comment]
> Seeing the Enketo fix is rolled out by the time we've got Build and build2xlsform PRs merged and deployed, do I understand you correctly to keep the orx prefixes?

Yes, that's what I think we should do. The issue won't affect data capture as far as I know, only edits. And there will be a solution -- upgrade Enketo/Central.

**Review by lognaturel — CHANGES_REQUESTED** [review]
I agree with the choices you've made on the UI/flow side. I think that to make this safer we should forget about namespacing meta and audit. You'll also need to make sure that the node name is fixed.

**Comment by florianm** [comment]
That makes sense and I'll send the changes.

Since the data name is now fixed too and we'll never want two audit fields, and the position in the form definition does not impact how audit gets represented in the XML, I'm wondering whether it wouldn't be best configured in the form settings near autosend/delete settings.

What is your preference, @lognaturel?

**Comment by florianm** [comment]
I suggest closing this PR in favour of #287 and #286.
